text/0000-wallet-replay-protection.md

```diff
@@ -0,0 +1,298 @@
+2. Call to `create` creates a specific `r` value for `Protected` outputs e.g. they should start with `N` zeros or be divisible by some number `M`
+3. Call to `create` uses additional output information in the ~30 bytes that are available in the Bulletproofs to convey the idea whether the output is protected. Perhaps we could define a specific structure for these bytes e.g. `<scheme_version:1 byte><meta_data:4 bytes><data:25 bytes>` where `metadata` would also tell whether the `data` that follows is encrypted or not. The data could be encrypted using the `seed` key and could thus hold information on output labels which would only be available to the owner of the output. We could even include the starting bytes of the anchor that protects it if we wanted to. If we went such path, we would need to think of the possible drawbacks.
+
+In all cases, the output label should only be visible to the owner of the output. It seems necessary to have labeling information about the UTXO on the UTXO itself if we want it to be consistent with different wallet reusing the same seed. If the information is held only on the wallet side, then we hit much bigger issues because there comes a need for either a manual intervention and labeling or a robust solution for synchronization between wallets - which does not appear simple to build.
+
+### Simple bootstrapping of protected outputs
+
+We start off with all of our outputs marked as `Unprotected`. To create a `Protected` output out of nothing we create the following outputs:
+1. an `anchor` output that has a form `0*H + r*G` - generated from key derivation path `A`
+2. a set of protected outputs `PS` that hold our coins - generated from key derivation path `P`
+
+We make a bootstrap transaction with `anchor` and `PS` outputs. The outputs in `PS` are exceptionally labeled as `Protected` because they are joined in the same transaction output set as our anchor output. This transactions cannot be replayed because the anchor output will never be spent which means that `PS` outputs are safe from being recreated.
+
+_Note: Grin has a rule that an output that already exists in the UTXO set cannot be created which makes the transaction invalid._
+
+### Wallet transaction rules
+
+There are two known ways to protect against replay attacks using wallet rules. One is to utilize the `Protected` outputs in transactions, the other one is to keep a history of outputs that were spent and label as 'unsafe' those that the wallet does not remember creating.
+
+#### Replay protection with utilization of Protected outputs
+
+In order to protect ourselves from replay attacks, we need to follow a simple rule:
+**When we are sending money to someone , we _MUST_ always include a `Protected` input as a part of the transaction.** This is to prevent someone doing a replay of the transaction which would move money from our outputs. Exception to this rule are self-spends.

```

Comment:
Self-spends of unprotected outputs are still problematic. If recreated, they don't behave as other unprotected outputs.

Both protected and unprotected outputs come in two varieties. Signed-for, and not-signed-for. Note that I don't use spent and not-spent. If it's signed-for, it's still subject to play attacks, and should be treated with care.

Reply:
I'm not sure what you mean by "they don't behave as other unprotected outputs"? Let's say I receive an unprotected output `O1` from a 1-2 tx. My current understanding is that there are two ways to do a self spend which includes `O1`:
1. with another protected output - this makes the new outputs protected and hence safe
2. without a protected output e.g. using `O1` as an input and creating `O2` and `O3` outputs that are not protected
From my understanding, in the 2. case, `O2` and `O3` have the same properties with regards to `replay` attacks as `O1`. It's not different from receiving a 1-3 tx with `O2` and `O3` being unprotected. I think I didn't make the scenario I described in this comment clear in the document and assumed it was obvious (which it isn't), right? Or did I miss something else?

Regarding your second point, I agree, signed and unsigned is the correct way to think about it to protect from the `play` attacks - I put the play attacks towards in the end of the document.